crates/signer/src/signer.rs

```diff
@@ -0,0 +1,214 @@
+use crate::{Result, Signature};
+use alloy_primitives::{eip191_hash_message, Address, B256};
+use async_trait::async_trait;
+
+#[cfg(feature = "eip712")]
+use alloy_sol_types::{Eip712Domain, SolStruct};
+
+/// Asynchronous Ethereum signer.
+///
+/// All provided implementations rely on [`sign_hash`](Signer::sign_hash). A signer may not always
+/// be able to implement this method, in which case it should return
+/// [`UnsupportedOperation`](crate::Error::UnsupportedOperation), and implement all the signing
+/// methods directly.
+///
+/// Synchronous signers should implement both this trait and [`SignerSync`].
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait Signer: Send + Sync {
```

Comment:
I think we want to impl this for `Box` as well?
but can do seprately

Reply:
can do `&mut` and `Box` on Signer only due to `set_chain_id`